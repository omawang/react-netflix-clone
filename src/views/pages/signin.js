import React, { useContext, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { FirebaseContext } from '../../contexts/firebase';
import { Form } from '../components';
import { FooterContainer } from '../containers/footer';
import { HeaderContainer } from '../containers/header';

import * as ROUTES from '../../constants/routes';

export default function Signin() {
  const history = useHistory();
  const { firebase } = useContext(FirebaseContext);

  const [emailAddress, setEmailAddress] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState('');

  const isInvalid = emailAddress === '' || password === '';

  const handleSignIn = (e) => {
    e.preventDefault();
    firebase
      .auth()
      .signInWithEmailAndPassword(emailAddress, password)
      .then(() => {
        history.push(ROUTES.BROWSE);
      })
      .catch((error) => {
        setEmailAddress('');
        setPassword('');
        setError(error.message);
      });
  };

  return (
    <>
      <HeaderContainer>
        <Form>
          <Form.Title>Sign In Now</Form.Title>
          {error && <Form.Error data-testid="error">{error}</Form.Error>}

          <Form.Base method="POST" onSubmit={handleSignIn}>
            <Form.Input
              placeholder="Email address"
              value={emailAddress}
              onChange={({ target }) => setEmailAddress(target.value)}
            />

            <Form.Input
              placeholder="Password"
              type="password"
              autoComplete="off"
              value={password}
              onChange={({ target }) => setPassword(target.value)}
            />

            <Form.Submit
              disabled={isInvalid}
              type="submit"
              data-testid="signin-submit"
            >
              Sign In
            </Form.Submit>
          </Form.Base>

          <Form.Text>
            New to Netflix?{' '}
            <Form.Link to={ROUTES.SIGN_UP}>Sign up now.</Form.Link>
          </Form.Text>
          <Form.TextSmall>
            This page is protected by Google reCAPTCHA to ensure you&apos;re not
            a bot. Learn more.
          </Form.TextSmall>
        </Form>
      </HeaderContainer>

      <FooterContainer />
    </>
  );
}
