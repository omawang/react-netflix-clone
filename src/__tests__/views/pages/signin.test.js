import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { fireEvent, render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import { FirebaseContext } from '../../../contexts/firebase';
import { Signin } from '../../../views/pages';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({}),
}));

const firebase = {
  auth: () => ({
    signInWithEmailAndPassword: jest.fn(() => {
      return Promise.resolve('I am signed in!');
    }),
  }),
};

describe('Signin page', () => {
  it('render Signin page', async () => {
    const {
      container,
      queryByTestId,
      getByPlaceholderText,
      getByTestId,
    } = render(
      <FirebaseContext.Provider value={{ firebase }}>
        <Router>
          <Signin />
        </Router>
      </FirebaseContext.Provider>
    );

    await act(async () => {
      await fireEvent.change(getByPlaceholderText('Email address'), {
        target: { value: 'awank@app.com' },
      });
      await fireEvent.change(getByPlaceholderText('Password'), {
        target: { value: 'secret' },
      });
      fireEvent.click(getByTestId('signin-submit'));

      expect(getByPlaceholderText('Email address').value).toBe('awank@app.com');
      expect(getByPlaceholderText('Password').value).toBe('secret');

      expect(queryByTestId('error')).toBeFalsy();
    });

    // expect(getByText('Sign In Now')).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
