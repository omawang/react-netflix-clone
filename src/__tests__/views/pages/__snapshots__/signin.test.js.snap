// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signin page render Signin page 1`] = `
<div
  class="header__Background-sc-1v45rsj-0 gvksrW"
>
  <div
    class="header__Container-sc-1v45rsj-1 eSeMKt"
  >
    <a
      href="/"
    >
      <img
        alt="Netflix"
        class="header__Logo-sc-1v45rsj-6 kbpjKE"
        data-testid="header-bg"
        src="logo.svg"
      />
    </a>
    <a
      class="header__ButtonLink-sc-1v45rsj-7 jNQYQi"
      href="/signin"
    >
      Sign In
    </a>
  </div>
  <div
    class="form__Container-sc-1ap38re-0 dplSrE"
  >
    <h1
      class="form__Title-sc-1ap38re-1 gksWve"
    >
      Sign In Now
    </h1>
    <div
      class="form__Error-sc-1ap38re-2 igbMtA"
      data-testid="error"
    >
      history.push is not a function
    </div>
    <form
      class="form__Base-sc-1ap38re-3 fYWkjo"
      method="POST"
    >
      <input
        class="form__Input-sc-1ap38re-7 muXbO"
        placeholder="Email address"
        value=""
      />
      <input
        autocomplete="off"
        class="form__Input-sc-1ap38re-7 muXbO"
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        class="form__Submit-sc-1ap38re-8 koHILv"
        data-testid="signin-submit"
        disabled=""
        type="submit"
      >
        Sign In
      </button>
    </form>
    <p
      class="form__Text-sc-1ap38re-4 gvpEcr"
    >
      New to Netflix?
       
      <a
        class="form__Link-sc-1ap38re-6 kyLsmz"
        href="/signup"
      >
        Sign up now.
      </a>
    </p>
    <p
      class="form__TextSmall-sc-1ap38re-5 cSdqcd"
    >
      This page is protected by Google reCAPTCHA to ensure you're not a bot. Learn more.
    </p>
  </div>
</div>
`;
